File: leetcode/139. Word Break.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        max_len = max([len(word) for word in wordDict])
        word_set = set(wordDict)
        
        dp = [False] * (len(s)+1)
        dp[0] = True
        
        for i in range(len(s)):
            if not dp[i]:
                continue
            for j in range(i, min(len(s), i+max_len)):
                word = s[i:j+1]
                if word in word_set:
                    dp[j+1] = True

        return dp[-1]

File: leetcode/139. Word Break.kt
class Solution {
    fun wordBreak(s: String, wordDict: List<String>): Boolean {
        val lengths = wordDict.map { it.length }
        val min_len = lengths.min()
        val max_len = lengths.max()
        val wordSet = wordDict.toSet()

        val dp = Array(s.length + 1) { false }
        dp[0] = true
        for (i in 0..s.length-1) {
            if (!dp[i]) continue
            for (j in i+min_len..min(i+max_len, s.length)) {
                val sub = s.substring(i, j)
                if (sub in wordSet) {
                    dp[j] = true
                    if (j == s.length) return true
                }
            }
        }

        return false
    }
}
